### ISSUE TITLE
Check for basic usability with screen reader

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-flutter#535]
We should try out the app with iOS's VoiceOver and/or Android's TalkBack, and make sure it's possible to do the essentials to be able to use the app at all:
* log in
* get between the different top-level pages (Inbox, All messages, Subscribed streams, Direct messages)
* use each of those to get to a relevant message list
* read the messages
* send a message

If there's anything that doesn't work, we should file and triage an issue for that so we can fix it.

Generally the fixes will involve Flutter's [`Semantics`](https://main-api.flutter.dev/flutter/widgets/Semantics-class.html) widget and related parts of the Flutter API, to provide the appropriate semantic information for screen readers (like VoiceOver and TalkBack) and other assistive technology to use.

### How to approach this issue

See my discussion in https://github.com/zulip/zulip-mobile/issues/4394 about how to start learning to use these systems and to evaluate the app with them.

### Not in scope

This issue is milestoned "Launch". Out of scope for this issue is to do any more comprehensive review of the rest of the app's features: reactions, less-common message content features, action sheets, lightbox, compose typeaheads, etc., etc. Those features are also important, and I'll file a companion issue to this one for covering all those, which will get the "Post-launch" milestone.

The current state in zulip-mobile — the last I looked, which I see was three years ago — is described in https://github.com/zulip/zulip-mobile/issues/4394: the app basically works, but there are a long list of glitches. We did fix the two issues highlighted as most important (in https://github.com/zulip/zulip-mobile/issues/4232) by an organization with some people who actually use screen readers regularly; but we don't really know how usable the app is for such users.

**Comment by MINEGHOST007** [comment]
I would like to work on this issue I'll be able to provide screenshots and recordings for Android TalkBack .

**Comment by gnprice** [comment]
Sure, please go ahead.

Recordings will be a helpful bonus, but please also include transcriptions, along with those screenshots for context. The transcriptions will help make it easy for others to review your findings.

For detailed discussion, please start a chat thread on chat.zulip.org in #mobile-dev-help. (Zulip makes a better medium for discussion than GitHub does.) Then after we have some concrete findings we're confident in, we can post those here on this issue thread, keeping this thread easy to read.

**Comment by AshutoshKhadse23** [comment]
Hello @gnprice ,
Is someone working on it already if not then I would like to work on it ?
I seen the comment but unable to find the updates on that CZO and here.

**Comment by gnprice** [comment]
To find chat threads related to a particular issue or PR, I recommend searching for the issue or PR's name as recognized by [the chat.zulip.org linkifiers](https://zulip.com/development-community/#linking-to-github-issues-and-pull-requests): so for this issue, "F535". That finds this thread: [#mobile-team > #F535 Android TalkBack Testing @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/.23F535.20Android.20TalkBack.20Testing/near/2141825)

I think this is an issue that would benefit from multiple people working on it independently, though. So @AshutoshKhadse23 please go ahead with your own investigation. If you're able to test the behavior on iOS with VoiceOver, that would be especially helpful (because @MINEGHOST007's testing was on Android with TalkBack); but if not, a second independent test on Android is also useful.

**Comment by PIG208** [comment]
This is a more boiled-down version of https://github.com/zulip/zulip-flutter/issues/537#issuecomment-2973273681 containing things that I found more relevant for early post-launch:

[edit 2025-10-31 by gnprice: linked to newly-filed issues]

- [ ] The home page bottom tabs do not have labels. The caption for each of them is just: "button, double tap to activate". → #1960
- [ ] The "----OR-----" separator on the login page might need a caption that elaborates on what it means. Without a visual cue, it is a bit abrupt as a single word. (Note, by navigating the third-party sign-in sites, we can compare those pages' accessibility with ours quite conveniently.) Trying it on GitHub, there is an invisible element right after the "sign in" button! It reads "password log-in alternatives", and the option after it is "sign in with a passkey". → #1961
- [ ] The loading screen/loading indicator cannot get focused. There should at least be some way to tell if the page is loading. → #1962
- [ ] Message content navigation appears out of order when the message contains links. → #1963
- [ ] Headers should be marked as such. This includes recipient headers in the message list and the collapsible channel headings on the inbox page. This is probably quite important for the inbox page, since there can be difficulty telling channel and topic items apart. → #1964
- [ ] Navigating backwards (to older messages) does not work smoothly. Sometimes the focus can get moved to the sticky header instead of the previous message; sometimes the focus moves off-screen or behind the sticky header, instead of automatically scrolled. For reference on the correct behavior, see what happens when you are about to navigate to a channel item on the channels page off-screen. → #1965
- [ ] Autocomplete options not disappearing after the keyboard is collapsed might obstruct the message list, and the user might not be aware of that. This is #1098.

**Comment by emilysutherlin** [comment]
A user in a large open-source community uses a screenreader and observes that the mobile app is difficult to navigate due to several unlabeled tabs and controls.

**Comment by gnprice** [comment]
I think the "check" step is complete, thanks to @PIG208 at https://github.com/zulip/zulip-flutter/issues/535#issuecomment-2973276995 above. Those results include some unlabeled tabs and controls, which are likely candidates for the ones the user mentioned at https://github.com/zulip/zulip-flutter/issues/535#issuecomment-3207350488 was running into.

So I think the next steps to do here are:
* File an issue for each of the items in https://github.com/zulip/zulip-flutter/issues/535#issuecomment-2973276995 .
* Close this issue.
* Fix those specific issues one by one 🙂. Maybe start with the ones that are about unlabeled tabs and controls: the home page bottom tabs, and headers in the message list and the inbox.

**Comment by gnprice** [comment]
OK, I've filed the following 6 issues:
* #1960
* #1961
* #1962
* #1963
* #1964
* #1965

and added a note on one existing issue:
* #1098

all reflecting the findings reported at https://github.com/zulip/zulip-flutter/issues/535#issuecomment-2973276995 above.

With that, I think this issue is complete. Thanks again @pig208!
